File: app/src/main/java/com/example/earthwallet/bridge/models/Permit.java
package com.example.earthwallet.bridge.models;

import java.util.List;

/**
 * SNIP-24 Permit data model
 *
 * Represents a query permit that allows authentication without requiring
 * a prior transaction to set viewing keys.
 */
public class Permit {

    private String permitName;
    private List<String> allowedTokens;
    private List<String> permissions;
    private String signature;
    private String publicKey;
    private long timestamp;

    public Permit() {}

    public Permit(String permitName, List<String> allowedTokens, List<String> permissions) {
        this.permitName = permitName;
        this.allowedTokens = allowedTokens;
        this.permissions = permissions;
        this.timestamp = System.currentTimeMillis();
    }

    public String getPermitName() {
        return permitName;
    }

    public void setPermitName(String permitName) {
        this.permitName = permitName;
    }

    public List<String> getAllowedTokens() {
        return allowedTokens;
    }

    public void setAllowedTokens(List<String> allowedTokens) {
        this.allowedTokens = allowedTokens;
    }

    public List<String> getPermissions() {
        return permissions;
    }

    public void setPermissions(List<String> permissions) {
        this.permissions = permissions;
    }

    public String getSignature() {
        return signature;
    }

    public void setSignature(String signature) {
        this.signature = signature;
    }

    public String getPublicKey() {
        return publicKey;
    }

    public void setPublicKey(String publicKey) {
        this.publicKey = publicKey;
    }

    public long getTimestamp() {
        return timestamp;
    }

    public void setTimestamp(long timestamp) {
        this.timestamp = timestamp;
    }

    /**
     * Check if permit contains permission for a specific action
     */
    public boolean hasPermission(String permission) {
        return permissions != null && permissions.contains(permission);
    }

    /**
     * Check if permit allows queries for a specific token contract
     */
    public boolean allowsToken(String contractAddress) {
        return allowedTokens != null && allowedTokens.contains(contractAddress);
    }

    /**
     * Check if permit is still valid (basic validation)
     */
    public boolean isValid() {
        return permitName != null && !permitName.isEmpty() &&
               allowedTokens != null && !allowedTokens.isEmpty() &&
               permissions != null && !permissions.isEmpty() &&
               signature != null && !signature.isEmpty() &&
               publicKey != null && !publicKey.isEmpty();
    }
}

File: app/src/main/java/com/example/earthwallet/bridge/models/Permit.kt
package network.erth.wallet.bridge.models

/**
 * SNIP-24 Permit data model
 *
 * Represents a query permit that allows authentication without requiring
 * a prior transaction to set viewing keys.
 */
data class Permit(
    var permitName: String? = null,
    var allowedTokens: List<String>? = null,
    var permissions: List<String>? = null,
    var signature: String? = null,
    var publicKey: String? = null,
    var timestamp: Long = System.currentTimeMillis()
) {

    /**
     * Constructor with required fields
     */
    constructor(
        permitName: String,
        allowedTokens: List<String>,
        permissions: List<String>
    ) : this(
        permitName = permitName,
        allowedTokens = allowedTokens,
        permissions = permissions,
        timestamp = System.currentTimeMillis()
    )

    /**
     * Check if permit contains permission for a specific action
     */
    fun hasPermission(permission: String): Boolean {
        return permissions?.contains(permission) ?: false
    }

    /**
     * Check if permit allows queries for a specific token contract
     */
    fun allowsToken(contractAddress: String): Boolean {
        return allowedTokens?.contains(contractAddress) ?: false
    }

    /**
     * Check if permit is still valid (basic validation)
     */
    fun isValid(): Boolean {
        return !permitName.isNullOrEmpty() &&
               !allowedTokens.isNullOrEmpty() &&
               !permissions.isNullOrEmpty() &&
               !signature.isNullOrEmpty() &&
               !publicKey.isNullOrEmpty()
    }
}